draft-pda-protocol.md

```diff
@@ -573,116 +612,288 @@ PA protocol. The structure of the leader share is similarly protocol specific:
+checks that checks that `PAAggregateSubReq.proto == PAParam.proto`. If not, it
+aborts and alerts the leader with "incorrect protocol for sub-request".
+Otherwise, It computes the HPKE context as
 
 ~~~
-context = SetupBaseR(PAVerifyReq.enc, sk, [TODO])
+context = SetupBaseR(PAAggregateSubReq.enc, sk, [TODO])
 ~~~
 
-where `sk` is the secret key corresponding to the HPKE config. Next, it computes
-the body of the `PAVerifyResp` according to the PA protocol.
+where `sk` is its HPKE key and computes the body of the `PAAggregateSubResp`. It
+then updates its state according to the PA protocol. After processing all of the
+sub-requests, the helper encrypts its updated state and constructs its response
+to the aggregate request.
 
-[OPEN ISSUE: encrypt and store Helper state at the Leader]
+##### Helper State
 
-## Collect {#pa-collect}
+The helper state is an optional parameter of an aggregate request that the can
+helper use to carry state across requests. At least part of the state will
+usually need to be encrypted in order to protect user privacy. However, the
+details of precisely how the state is encrypted and the information that it
+carries is up to the helper implementation.
+
+#### Output Share Request
+
+Once the aggregators have verified at least as many reports as required for the
+PA task, the leader issues an *output share request* to the helper. The helper
+responds to this request by extracting its output share from its state and
+encrypting it under the collector's HPKE public key.
+
+The leader sends a POST request to `[helper]/output_share` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+  uint64 batch_start; // Same as PACollectReq.batch_start.
+  Uint64 batch_end;   // Same as PACollectReq.batch_end.
+  opaque helper_state<0..2^16>;
+} PAOutputShareReq;
+~~~
+
+To respond to valid output share requests, the helper first checks that
+`batch_start` and `batch_end` are multiples of `task.batch_window` and that
+`batch_end - batch_start >= task.batch_window`. Next, it extracts from its state
+the set of input shares that fall in the window `[batch_start, batch_end)`. If
+the size of the batch is less than `task.batch_size`, then it aborts and alerts
+the leader with "insufficient data". Otherwise, it computes its output share,
+which has the following structure:
+
+~~~
+struct {
+  PAProto proto;
+  select (PAOutputShare.proto) {
+    case prio: PrioOutputShare;
+    case hits: HitsOutputShare;
+  }
+} PAOutputShare;
+~~~
 
-[TODO]
+Next, it encrypts its output share under the collector's HPKE public key:
+
+~~~
+enc, context = SetupBaseS(pk, [TODO])
+encrypted_output_share = context.Seal(output_share, [TODO])
+~~~
+
+where `pk` is the HPKE public key encoded by the collector's HPKE key
+configuration and `output_share` is its serialized output share.
+
+It responds with the following message:
+
+~~~
+struct {
+  uint8 collector_hpke_config_id;
```

Comment:
I don't see how this works and why is it just one byte?

Reply:
This is basically the same thing we do in ECH: there is a 1-byte config identifier that's used to inform the decrypting party which HPKE key it needs to decrypt. It's a byte only because other HPKE applications have only a byte for the config id.